modules/telemetry/telemetry/include/hephaestus/telemetry/struclog.h

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <iomanip>
+#include <source_location>
+#include <sstream>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <utility>
+#include <vector>
+
+namespace heph::telemetry {
+
+enum class Level : std::uint8_t { TRACE, DEBUG, INFO, WARN, ERROR, FATAL };
+auto operator<<(std::ostream& /*os*/, const Level& /*level*/) -> std::ostream&;
+
+template <typename T>
+concept NotQuotable = !std::convertible_to<T, std::string> && requires(std::ostream& os, T t) {
+  { os << t } -> std::same_as<std::ostream&>;
+};
+
+template <typename T>
+concept Quotable = std::convertible_to<T, std::string>;
+
+template <typename T>
+struct Field final {
+  std::string key;
+  T val;
+};
+
+/**
+ * @brief A class that allows easy composition of logs for structured logging.
+ *        Example(see also struclog.cpp):
+ *        namespace ht=heph::telemetry;
+ *        log(ht::LogEntry{Level::INFO, "my module", "adding"} | "id"_f(12345) | ht::Field{"tag", "test"}) ;
+ *
+ *        logs
+ *        'level=info hostname=goofy location=struclog.h:123 thread-id=5124 time=2023-12-3T8:52:02+0
+ * module="my module" message="adding" id=12345 tag="test"'
+ *
+ *        Remark: We would like to use libfmt with quoted formatting as proposed in
+ *                https://github.com/fmtlib/fmt/issues/825 once its included (instead of sstream).
+ *
+ */
+struct LogEntry {
+  using Fields = std::vector<Field<std::string>>;
+  using Clock = std::chrono::system_clock;
```

Comment:
yes it is complicated and good to be modular, however for now I think it is sufficient. I am usually careful to add variability and generalization if not explicitely needed... since the log time usually is interesting up to frame rate, i guess the resolution of this clock is fine for us now an I would add the generalization if needed.

Reply:
I agree, logs are not metric or profiles, so we don't have a strong time resolution need

Reply:
This point goes back to the discussion with @floriantschopp . From what I understood, Voliro has the exact issue of logs becoming async due to drifting clocks. Maybe @floriantschopp can wager in here.

Reply:
added an issue for that: https://github.com/olympus-robotics/hephaestus/issues/208